src/pfe/portal/routes/projects/links.route.js

```diff
@@ -0,0 +1,221 @@
+      // Get the url from the projectList, throw an error if the project does not exist
+      const localProjectURL = await verifyProjectExistsAndReturnInternalURL(user, targetProjectID);
+      newLink.projectURL = localProjectURL;
+      newLink.type = TYPES.LOCAL;
+    }
+
+    await project.createLink(newLink);
+    log.info(`New project link created for ${project.name}`);
+
+    // Send status and then kick off the restart/rebuild
+    const { links } = project;
+    res.status(200).send(links.getAll());
+
+    if (project.isOpen()) {
+      await restartOrBuildProject(user, project);
+    }
+  } catch (err) {
+    log.error(err);
+    if (err.code === ProjectLinkError.CODES.INVALID_PARAMETERS) {
+      res.sendStatus(400);
+    } else if (err.code === ProjectLinkError.CODES.EXISTS) {
+      res.sendStatus(409);
+    } else {
+      res.status(500).send(err);
+    }
+  }
+});
+
+router.put('/api/v1/projects/:id/links', validateReq, checkProjectExists, async(req, res) => {
+  const currentEnvName = req.sanitizeBody('envName');
+  const newEnvName = req.sanitizeBody('updatedEnvName');
+  let newProjectURL = req.sanitizeBody('targetProjectURL');
+  try {
+    const { cw_user: user } = req;
+    const project = getProjectFromReq(req);
+    const { links } = project;
+
+    // If the link on the same PFE (local) fetch the projectURL from the ProjectList (ignore newProjectURL)
+    const { type: linkType, projectID: targetProjectID, projectURL: currentProjectURL } = links.get(currentEnvName);
+    
+    if (linkType === TYPES.LOCAL) {
+      // Get the url from the projectList, throw an error if the project does not exist
+      const localProjectURL = await verifyProjectExistsAndReturnInternalURL(user, targetProjectID);
+      newProjectURL = localProjectURL;
+    }
+
+    // If newEnvName or newProjectURL are not given through the API, default them to their old values
+    const updatedLinkEnvName = (newEnvName) ? newEnvName : currentEnvName;
+    const updatedLinkProjectURL = (newProjectURL) ? newProjectURL : currentProjectURL;
+    await project.updateLink(currentEnvName, updatedLinkEnvName, updatedLinkProjectURL);
+
+    // Send status and then kick off the restart/rebuild
+    res.sendStatus(204);
+
+    if (project.isOpen()) {
+      await restartOrBuildProject(user, project);
+    }
+  } catch (err) {
+    log.error(err);
+    if (err.code === ProjectLinkError.CODES.NOT_FOUND) {
+      res.sendStatus(404);
+    } else {
+      res.status(500).send(err);
+    }
+  }
+});
+
+router.delete('/api/v1/projects/:id/links', validateReq, checkProjectExists, async(req, res) => {
+  const envNameOfLinkToDelete = req.sanitizeBody('envName');
+  try {
+    const { cw_user: user } = req;
+    const project = getProjectFromReq(req);
+    await project.deleteLink(envNameOfLinkToDelete);
+
+    // Send status and then kick off the restart/rebuild
+    res.sendStatus(204);
+
+    if (project.isOpen()) {
+      await restartOrBuildProject(user, project);
```

Comment:
Can this throw an exception? (I think it can further down in the restart code.) If so that would result in calling `res.sendStatus()` again and we'd get the message we usually get about not being able to set headers after they have been sent.

Reply:
You're right, I'll change this